dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/mvc/interceptor/AuthorityInterceptor.java

```diff
@@ -0,0 +1,100 @@
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.webapi.mvc.interceptor;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import javax.annotation.Nonnull;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import lombok.extern.slf4j.Slf4j;
+import org.hisp.dhis.security.Authorities;
+import org.hisp.dhis.security.RequiresAuthority;
+import org.springframework.security.access.AccessDeniedException;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.security.core.context.SecurityContext;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.security.web.authentication.session.SessionAuthenticationException;
+import org.springframework.stereotype.Component;
+import org.springframework.web.method.HandlerMethod;
+import org.springframework.web.servlet.HandlerInterceptor;
+
+/**
+ * Interceptor which checks that the {@link org.hisp.dhis.user.User} has any one of {@link
+ * org.hisp.dhis.security.Authorities} which are passed in {@link RequiresAuthority} as an arg.
+ *
+ * <p>Throws {@link AccessDeniedException} if {@link org.hisp.dhis.user.User} does not have any of
+ * the passed-in {@link org.hisp.dhis.security.Authorities}.
+ *
+ * <p>{@link Authorities#ALL} is automatically added to the check, as having this Authority allows
+ * access to all methods by default.
+ */
+@Slf4j
+@Component
+public class AuthorityInterceptor implements HandlerInterceptor {
+
+  @Override
+  public boolean preHandle(
+      @Nonnull HttpServletRequest request,
+      @Nonnull HttpServletResponse response,
+      @Nonnull Object handler) {
+    HandlerMethod handlerMethod = (HandlerMethod) handler;
+
+    if (!handlerMethod.hasMethodAnnotation(RequiresAuthority.class)) {
+      return true;
+    }
+
+    RequiresAuthority requiresAuthority =
+        handlerMethod.getMethodAnnotation(RequiresAuthority.class);
+    if (requiresAuthority != null) {
+      // include 'ALL' authority in required authorities
+      List<Authorities> requiredAuthorities = new ArrayList<>(List.of(Authorities.ALL));
+      requiredAuthorities.addAll(List.of(requiresAuthority.anyOf()));
+
+      // get user authorities
+      final SecurityContext securityContext = SecurityContextHolder.getContext();
+      Authentication authentication = securityContext.getAuthentication();
+      if (authentication == null)
+        throw new SessionAuthenticationException("Error trying to get user authentication details");
```

Comment:
@netroms This kind of ties in to what we would like to do in future (instead of returning null for auth errors). Does this look ok to you? I wanted to call this out so you're aware of it, to make sure we're dealing with these errors in a consistent manner.
Returning null is not an option in this part of the code, it requires a boolean answer & true (has required auth) or false (does not have required auth, we don't know as it may be a system issue?) are both invalid responses in this context.